src/Reader.cpp

```diff
@@ -0,0 +1,114 @@
+#include "Reader.h"
+
+Reader::Reader(SourcePtr source, TokenizerPtr tokenizer, std::vector<CollectorPtr> collectors,
+                    CharacterVector colNames, LocaleInfo* locale,
+                    int progress) :
+  source_(source),
+  tokenizer_(tokenizer),
+  collectors_(collectors),
+  locale_(locale),
+  progress_(progress) {
+
+  tokenizer_->tokenize(source_->begin(), source_->end());
+  tokenizer_->setWarnings(&warnings_);
+
+
+  // Work out which output columns we are keeping and set the locale for each collectors
+  size_t p = collectors_.size();
+  for (size_t j = 0; j < p; ++j) {
+    if (!collectors_[j]->skip()) {
+      keptColumns_.push_back(j);
+      collectors_[j]->setWarnings(&warnings_);
+    }
+  }
+
+    if (colNames.size() > 0) {
+      outNames_ = CharacterVector(keptColumns_.size());
+      int i = 0;
+      for (std::vector<int>::const_iterator it = keptColumns_.begin(); it != keptColumns_.end(); ++it) {
+        outNames_[i++] = colNames[*it];
+      }
+    }
+}
+
+RObject Reader::readToDataFrame(int lines) {
+  int rows = read(lines);
+
+  // Save individual columns into a data frame
+  List out(outNames_.size());
+  int j = 0;
+  for (std::vector<int>::const_iterator it = keptColumns_.begin(); it != keptColumns_.end(); ++it) {
+    out[j++] = collectors_[*it]->vector();
+  }
+
+  out.attr("class") = CharacterVector::create("tbl_df", "tbl", "data.frame");
+  out.attr("row.names") = IntegerVector::create(NA_INTEGER, -(rows + 1));
+  out.attr("names") = outNames_;
+
+  return warnings_.addAsAttribute(out);
+}
+
+int Reader::read(int lines) {
+
+  int n = (lines < 0) ? 1000 : lines;
+
+  collectorsResize(collectors_, n);
+
+  int last_row = -1, last_col = -1, cells = 0;
+  for (Token t = tokenizer_->nextToken(); t.type() != TOKEN_EOF; t = tokenizer_->nextToken()) {
+    if (progress_ && (cells++) % progressStep_ == 0) {
```

Comment:
I had it as a variable because the value was different between read_lines_ and read_tokens_ (25,000 vs 250,000 IIRC), but I am happy to keep it constant!